### PR TITLE
I2C fix for MCP23017 / Adafruit RGBLCDShield

### PR DESCRIPTION
**Pull request by m-mcgowan**
contribution from @pra in the community. Thanks! https://community.particle.io/t/i2c-lcd-freezes-photon/15666

If you have a MCP23017 or Adafruit RGBLCDShield please test and report if this improves I2C connectivity!

### CODE REVIEW

**Comment by dresslerc** [comment]
I've been using the fix on 2 MCP23017 for 24 hours and they are still running strong.

My test code would usually freeze the i2c bus within 10-30 mins and zero issues so far.  I'll keep testing and report back after it runs for a couple days.

**Comment by pomplesiegel** [comment]
The MCP23017 locking up SDA was the original reason we were testing the fix `Wire.reset()` #598. Hopefully this takes things to the next level, preventing the problem before it occurs.

File: hal/src/stm32f2xx/i2c_hal.c [hunk 1]
```diff
@@ -1,726 +1,740 @@
...
+            /* SW Reset the I2C Peripheral */
+            HAL_I2C_SoftwareReset(i2c);
+
+            return 0;
+        }
+    }
+
+    /* perform blocking read into buffer */
+    uint8_t *pBuffer = i2cMap[i2c]->rxBuffer;
+    uint8_t numByteToRead = quantity;
```
Comment (technobly) [inline hal/src/stm32f2xx/i2c_hal.c]: These three lines and comment are disconnected from the code they are used with on line 315.  Doesn't look like they needed to move up here.

File: hal/src/stm32f2xx/i2c_hal.c [hunk 2]
```diff
@@ -1,726 +1,740 @@
...
+
+	if (quantity == 1)
+		/* Disable Acknowledgement */
+		I2C_AcknowledgeConfig(i2cMap[i2c]->I2C_Peripheral, DISABLE);
```
Comment (technobly) [inline hal/src/stm32f2xx/i2c_hal.c]: Disabling ACK if quantity == 1 here (no comment as to why that is needed before Slave Address send), but disabling ACK if numByteToRead (aka quantity) == 2 below later on receive.  Can we explain this difference better in comments?

File: hal/src/stm32f2xx/i2c_hal.c [hunk 3]
```diff
@@ -1,726 +1,740 @@
...
+        }
+    }
+
+
+    /* While there is data to be read */
+    while(numByteToRead)
+    {
+        /* Wait for data register to be full */
+		/* This means the byte in the RX has already been acked */
+		/* Ack is sent on the xfer of teh byte to RX from the shift register */
```
Comment (technobly) [inline hal/src/stm32f2xx/i2c_hal.c]: These comments are very confusing.  I want to fix them, but do not fully understand yet what was changed here.

File: hal/src/stm32f2xx/i2c_hal.c [hunk 4]
```diff
@@ -1,726 +1,740 @@
...
+                        I2C_GenerateSTOP(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+                break;
+
+            default:
+                break;
+        }
+
+        /* Read the byte from the Slave */
+		/* This will enhable the transfer of the next byte (if any) from the shift register to the RX */
+		/* and also sending ACK/NACK for that byte */
```
Comment (technobly) [inline hal/src/stm32f2xx/i2c_hal.c]: These comments are easier to understand, but some reference to numByteToRead >2, == 2 and == 1 in context would be helpful.

**Comment by technobly** [comment]
Not trying to nitpick this change which is obviously working for users, just trying to make sure this fairly  complicated peripheral gets documented adequately as we make improvements ;-)  Thus ensuring the changes are widely understood and stay fixed as the code evolves.

**Pull request merged**